{"cells":[{"cell_type":"markdown","id":"69755b57-9eee-4b3b-8dd4-6140486a4182","metadata":{"tags":[]},"source":["# Initialize Notebook"]},{"cell_type":"markdown","id":"b11a388c-5ca1-4097-91a5-29e8dc6bd559","metadata":{"tags":[]},"source":["## Import packages"]},{"cell_type":"code","execution_count":1,"id":"cd14b439-3a34-4910-96d6-fc7afaede06f","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["import pyspark\n","import dxdata\n","import dxpy\n","import pandas as pd\n","import json\n","import requests\n","import time\n","import subprocess\n","import os\n","\n","sc = pyspark.SparkContext()\n","spark = pyspark.sql.SparkSession(sc)\n"]},{"cell_type":"markdown","id":"b4f97b50-339a-415e-8dbc-6fb2a896fc28","metadata":{},"source":["## Initialize helper functions"]},{"cell_type":"code","execution_count":2,"id":"55fd1cd9-ddd7-476a-ac8b-66055a6c5f7a","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def fetch_gene_info_ensembl(gene_names, species=\"human\", genome_version=\"GRCh38\", batch_size=50):\n","    gene_info_list = []\n","    server = \"https://rest.ensembl.org\"\n","    endpoint = f\"/lookup/symbol/{species}\"\n","    headers = {\"Content-Type\": \"application/json\"}\n"," \n","    def make_request(batch):\n","        response = requests.post(server + endpoint, headers=headers, json={\"symbols\": batch})\n","        if response.ok:\n","            return response.json()\n","        else:\n","            print(f\"Fetching failed for batch: {batch}, Status Code: {response.status_code}\")\n","            return {}\n"," \n","    for i in range(0, len(gene_names), batch_size):\n","        print(f\"Processing batch {i // batch_size + 1} / {len(gene_names) // batch_size + 1}\")\n","        batch = gene_names[i:i + batch_size]\n","        data = make_request(batch)\n"," \n","        for gene_name, gene_data in data.items():\n","            if \"seq_region_name\" in gene_data:\n","                gene_info_list.append({\n","                    \"gene_name\": gene_data.get(\"display_name\", gene_name),\n","                    \"chromosome\": gene_data[\"seq_region_name\"],\n","                    \"start\": int(gene_data[\"start\"]),\n","                    \"end\": int(gene_data[\"end\"]),\n","                    \"genome_version\": genome_version\n","                })\n","        time.sleep(0.5)\n","    return gene_info_list\n"]},{"cell_type":"markdown","id":"7bd58ec6-affa-42d2-8431-a43064705769","metadata":{},"source":["## Initialize variables"]},{"cell_type":"code","execution_count":3,"id":"ef11f5cb-29fd-489b-8593-e353f883eb2f","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["results_dir = \"/results\"\n","\n","! dx find projects --name \"chorea_wgs\" > projectid.txt\n","projectid = open(\"projectid.txt\", \"r\")\n","projectid = projectid.read()\n","projectid = projectid.split(\" : \")[0]\n"]},{"cell_type":"markdown","id":"07ab455d-30f9-4d03-9e71-a5f3f98ea79a","metadata":{"tags":[]},"source":["## Download files we need"]},{"cell_type":"code","execution_count":4,"id":"79ee254f-7c0d-4c9d-94f7-f6683771dc9b","metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[2K[===========================================================>] Completed 9,741,159 of 9,741,159 bytes (100%) /opt/notebooks/ukbb_imputed_genotypes_umap_linearsvc_predicted_labels.txtxt\n","\u001b[2K[===========================================================>] Completed 10,052 of 10,052 bytes (100%) /opt/notebooks/chorea_gene_names.txtxt\n"]}],"source":["! dx download /data/ukbb_imputed_genotypes_umap_linearsvc_predicted_labels.txt --overwrite\n","! dx download /data/chorea_gene_names.txt --overwrite\n"]},{"cell_type":"markdown","id":"5d89578b-908e-4575-befb-65b84a438804","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["## Fetch gene info"]},{"cell_type":"code","execution_count":null,"id":"1b6e7509-6eec-4fa3-b424-0d638307d064","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["with open(\"chorea_gene_names.txt\", \"r\") as file:\n","    gene_list = file.read().splitlines()\n","\n","gene_info_list = fetch_gene_info_ensembl(gene_names=gene_list)\n","df_gene_info = pd.DataFrame(gene_info_list)\n","df_gene_info[\"chromosome\"] = pd.Categorical(\n","    df_gene_info[\"chromosome\"],\n","    categories=[str(i) for i in range(1, 23)] + [\"X\", \"Y\", \"MT\"],\n","    ordered=True,\n",")\n","df_gene_info = df_gene_info.sort_values(by=[\"chromosome\", \"start\"]).reset_index(drop=True)\n"]},{"cell_type":"markdown","id":"eb765857-d145-4199-9f27-dc41c88e415d","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["## Combine overlapping genes"]},{"cell_type":"code","execution_count":null,"id":"18085085-23c6-4a1c-8bc7-99d8b674839d","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["merged = []\n","current = df_gene_info.iloc[0].copy()\n","for i in range(1, len(df_gene_info)):\n","    row = df_gene_info.iloc[i]\n","    if current[\"chromosome\"] == row[\"chromosome\"] and current[\"end\"] >= row[\"start\"]:\n","        current[\"end\"] = max(current[\"end\"], row[\"end\"])\n","        current[\"start\"] = min(current[\"start\"], row[\"start\"])\n","        current[\"gene_name\"] += \" + \" + row[\"gene_name\"]\n","    else:\n","        merged.append(current)\n","        current = row.copy()\n","\n","merged.append(current)\n","df_gene_info = pd.DataFrame(merged)\n","display(df_gene_info)\n"]},{"cell_type":"markdown","id":"9b0365e7-deaf-4fde-9e78-96fe772006ec","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["## Combine rows with overlapping WGS files"]},{"cell_type":"code","execution_count":null,"id":"ab4cd4dd-c750-407c-af91-9d379a5c31ab","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["df_b_vals = df_gene_info.copy()\n","df_b_vals.loc[:, \"b_start\"] = df_b_vals[\"start\"] // 20000 - 1\n","df_b_vals.loc[:, \"b_end\"] = df_b_vals[\"end\"] // 20000 + 2\n","df_b_vals = df_b_vals[[\"chromosome\", \"b_start\", \"b_end\"]]\n","\n","merged = []\n","current = df_b_vals.iloc[0].copy()\n","for i in range(1, len(df_b_vals)):\n","    row = df_b_vals.iloc[i]\n","    if current[\"chromosome\"] == row[\"chromosome\"] and current[\"b_end\"] >= row[\"b_start\"]:\n","        current[\"b_end\"] = max(current[\"b_end\"], row[\"b_end\"])\n","        current[\"b_start\"] = min(current[\"b_start\"], row[\"b_start\"])\n","    else:\n","        merged.append(current)\n","        current = row.copy()\n","\n","merged.append(current)\n","df_b_vals = pd.DataFrame(merged)\n","display(df_b_vals)\n"]},{"cell_type":"markdown","id":"e0c12741-b68b-45a4-8292-29917425a0ae","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["## Save gene info"]},{"cell_type":"code","execution_count":null,"id":"254937c3-17ae-4657-a7bf-2bbd8157bb3c","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["df_gene_info.to_csv(\"gene_info.txt\", index=False, sep=\"\\t\")\n","df_b_vals.to_csv(\"b_val_ranges.txt\", index=False, sep=\"\\t\")\n","\n","! dx upload gene_info.txt --path /results/gene_info.txt\n","! dx upload b_val_ranges.txt --path /results/b_val_ranges.txt\n"]},{"cell_type":"markdown","id":"6db4eb5c-7b62-4cca-9413-3f60389debfa","metadata":{"tags":[]},"source":["# Fetch cohorts"]},{"cell_type":"markdown","id":"9d22733d-d6b7-495e-9a22-0302b36e8800","metadata":{},"source":["## Grab participant data"]},{"cell_type":"code","execution_count":5,"id":"2bf34e68-fd2d-4555-a6f1-0ef893e46987","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["dispensed_dataset_id = dxpy.find_one_data_object(typename=\"Dataset\", name=\"app*.dataset\", folder=\"/\", name_mode=\"glob\")[\"id\"]\n","dataset = dxdata.load_dataset(id=dispensed_dataset_id)\n","participant = dataset[\"participant\"]\n"]},{"cell_type":"markdown","id":"9ea3457f-503d-4b56-9aed-43ca4255495d","metadata":{"tags":[]},"source":["## Retrieve Cases"]},{"cell_type":"code","execution_count":6,"id":"f917dc60-cc2c-4991-9732-5b632c39258f","metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stderr","output_type":"stream","text":["/opt/conda/lib/python3.11/site-packages/thrift/transport/TSSLSocket.py:53: DeprecationWarning: ssl.PROTOCOL_TLS is deprecated\n","  self._context = ssl.SSLContext(ssl_version)\n"]}],"source":["field_names = [\n","    \"eid\", \"p31\", \"p34\", \"p21022\", \"p131274\", \"p131012\", \"p40000_i0\",\n","    \"p131016\", \"p131018\", \"p131020\", \"p131022\", \"p131024\", \"p131026\", \n","    \"p131028\", \"p131030\", \"p131036\", \"p131038\", \"p131040\", \"p131042\", \n","    \"p131046\", \"p131056\", \"p131058\", \"p131062\", \"p131066\", \"p131068\", \n","    \"p131070\", \"p131074\", \"p131076\", \"p131078\", \"p131080\", \"p131082\", \n","    \"p131084\", \"p131086\", \"p131088\", \"p131090\", \"p131092\", \"p131094\", \n","    \"p131096\", \"p131098\", \"p131100\", \"p131102\", \"p131104\", \"p131106\", \n","    \"p131108\", \"p131110\", \"p131112\", \"p131114\", \"p131116\", \"p131120\", \n","    \"p131122\", \"p131124\", \"p131126\", \"p42018\", \"p42020\", \"p42022\", \n","    \"p42024\", \"p42028\", \"p42030\", \"p42032\", \"p42034\", \"p42036\", \n","]\n","df_cases = participant.retrieve_fields(names=field_names, coding_values=\"replace\", engine=dxdata.connect())\n","df_cases = df_cases.toPandas()\n"]},{"cell_type":"code","execution_count":7,"id":"f6709363-3dd9-4265-a3c0-324eab83c72f","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["df_cases.rename(columns={\n","    \"eid\":\"ID\",\n","    \"p31\":\"GENETIC_SEX\", \n","    \"p34\":\"BIRTH_YEAR\", \n","    \"p21022\":\"AGE_OF_RECRUIT\",\n","    \"p131274\":\"CHOREA_DATE\",\n","    \"p131012\":\"HUNTINGTON_DATE\",\n","    \"p40000_i0\":\"DATE_OF_DEATH\",\n","}, inplace=True)\n","df_cases[\"ID\"] = pd.to_numeric(df_cases[\"ID\"])\n"]},{"cell_type":"code","execution_count":8,"id":"4fe6d7fc-01b8-401f-a7f4-584ca57cdeef","metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stderr","output_type":"stream","text":["/tmp/ipykernel_1191/3094394521.py:2: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_chorea[\"AGE\"] = pd.to_datetime(df_chorea[\"CHOREA_DATE\"]).dt.year - df_chorea[\"BIRTH_YEAR\"]\n"]}],"source":["df_chorea = df_cases[~df_cases[f\"CHOREA_DATE\"].isna()]\n","df_chorea[\"AGE\"] = pd.to_datetime(df_chorea[\"CHOREA_DATE\"]).dt.year - df_chorea[\"BIRTH_YEAR\"]\n","\n","df_ancestries = pd.read_csv(\"ukbb_imputed_genotypes_umap_linearsvc_predicted_labels.txt\", sep=\"\\t\")\n","df_ancestries.rename(columns={\"label\":\"ancestry\", \"IID\":\"ID\"}, inplace=True)\n","df_ancestries = df_ancestries[[\"ID\",\"ancestry\"]]\n","df_chorea = df_chorea.merge(df_ancestries, on=\"ID\")\n","\n","ids_chorea = df_chorea[\"ID\"].tolist()\n"]},{"cell_type":"markdown","id":"e703dfb9-c45f-4f4e-b7c1-5371fcd1f38e","metadata":{"tags":[]},"source":["## Check for related individuals"]},{"cell_type":"code","execution_count":9,"id":"3d9514f1-6ea1-498a-9442-75b43e5b302c","metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[2K[===========================================================>] Completed 3,805,958 of 3,805,958 bytes (100%) /opt/notebooks/ukb_rel.datat\n","There are NO related individuals!\n"]}],"source":["! dx download '/Bulk/Genotype\\ Results/Genotype\\ calls/ukb_rel.dat' --overwrite\n","df_full_related = pd.read_csv('ukb_rel.dat', sep = ' ')\n","df_full_related = df_full_related[df_full_related['Kinship'] > 0.0884]\n","\n","df_related_cohort = df_full_related.loc[df_full_related['ID1'].isin(ids_chorea) & df_full_related['ID2'].isin(ids_chorea)]\n","df_related_cohort.reset_index(drop=True, inplace=True)\n","print(f\"There are {'NO ' if len(df_related_cohort) == 0 else ''}related individuals!\")\n"]},{"cell_type":"markdown","id":"4384b677-27ec-47f6-813b-2280a0b7205e","metadata":{"tags":[]},"source":["## Save the IDs of each participant to a txt file"]},{"cell_type":"code","execution_count":null,"id":"ebac2b16-bb6d-4efe-a1e7-1341037867b6","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["with open('ids_pre_vcf.txt', 'w') as file:\n","    for iid in ids_chorea:\n","        file.write(f\"{iid}\\n\")\n","        "]},{"cell_type":"markdown","id":"0873ee2d-bfa5-4cc1-9deb-9eaafde85999","metadata":{"tags":[]},"source":["# Filter out participants without WGS data"]},{"cell_type":"markdown","id":"47616c8e-b182-41b8-8df7-ff9e15cdfb93","metadata":{"tags":[]},"source":["## Find participants without WGS data"]},{"cell_type":"code","execution_count":null,"id":"f6119e27-701e-4e7d-9ef3-c6c626643bd6","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["cmd = f\"dx run swiss-army-knife \"\n","cmd += f\"-iin='/Bulk/DRAGEN\\ WGS/DRAGEN\\ population\\ level\\ WGS\\ variants,\\ pVCF\\ format\\ [500k\\ release]/chr1/ukb24310_c1_b1_v1.vcf.gz' \"\n","cmd += f\"-iin='/Bulk/DRAGEN\\ WGS/DRAGEN\\ population\\ level\\ WGS\\ variants,\\ pVCF\\ format\\ [500k\\ release]/chr1/ukb24310_c1_b1_v1.vcf.gz.tbi' \"\n","cmd += f\"-icmd='bcftools query -l ukb24310_c1_b1_v1.vcf.gz > pvcf_full_ids.txt' \"\n","cmd += f\"--instance-type mem1_hdd1_v2_x2 \"\n","cmd += f\"--destination '{projectid}:{results_dir}'\"\n","\n","subprocess.run(\n","    cmd, \n","    shell=True, \n",")\n"]},{"cell_type":"code","execution_count":null,"id":"0ea2e900-da98-454e-ab67-95feb6fb8ae6","metadata":{},"outputs":[],"source":["\"\"\"\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------ PAUSE HERE UNTIL PREVIOUS STEP COMPLETES ------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","\"\"\""]},{"cell_type":"markdown","id":"3d3dabc1-41f1-4cc5-9cc3-c9ca23f627db","metadata":{"tags":[]},"source":["## Filter ID lists and clinical data to only include participants with WGS data"]},{"cell_type":"code","execution_count":10,"id":"0d7f0a68-4066-404c-91b1-648c96b8fbc6","metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[2K[===========================================================>] Completed 3,924,328 of 3,924,328 bytes (100%) /opt/notebooks/pvcf_full_ids.txtxt\n","grep: ids_pre_vcf.txt: No such file or directory\n"]}],"source":["! dx download {results_dir}/pvcf_full_ids.txt --overwrite\n","! grep -Fwf pvcf_full_ids.txt ids_pre_vcf.txt > filtered_sample_ids.txt\n","! dx upload filtered_sample_ids.txt --path {results_dir}/sample_ids.txt\n"]},{"cell_type":"code","execution_count":13,"id":"b39847be-cebd-4927-af73-a5f49b629ea9","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["df_chorea[df_chorea[\"ID\"].isin([1900895, 1048019, 5270799])].to_csv(\"participants_with_variants.txt\", sep=\"\\t\", index=False)\n"]},{"cell_type":"code","execution_count":null,"id":"80cae7a8-bebd-4a74-a4e4-e30aca4c578f","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["with open('filtered_sample_ids.txt', 'r') as file:\n","    ids_chorea = [int(line.strip()) for line in file]\n","df_chorea = df_chorea[df_chorea[\"ID\"].isin(ids_chorea)]\n","display(df_chorea)\n"]},{"cell_type":"markdown","id":"8a6900e9-4307-4e70-82ed-0618ac29f2a7","metadata":{},"source":["## Save final data table"]},{"cell_type":"code","execution_count":null,"id":"4498972b-b461-4f32-aec2-4e503cdfbf1c","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["df_chorea.to_csv(f'chorea_cases.txt', header=True, index=False, sep=\"\\t\")\n","! dx upload chorea_cases.txt --path {results_dir}/chorea_cases.txt\n"]},{"cell_type":"markdown","id":"cfc4bc7c-09fd-412d-88b2-9435b141e066","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["# Fetch pVCF chunks for each gene of interest"]},{"cell_type":"code","execution_count":null,"id":"3bf884ee-c4ef-4c7d-97c6-841a24d768d7","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["for _, row in df_b_vals.iterrows():\n","    start_bval = row[\"b_start\"]\n","    end_bval = row[\"b_end\"]\n","    chrom = row['chromosome']\n","    \n","    for b_val in range(start_bval, end_bval + 1):\n","        cmd = f\"dx run swiss-army-knife \"\n","        cmd += f\"-iin='/Bulk/DRAGEN\\ WGS/DRAGEN\\ population\\ level\\ WGS\\ variants,\\ pVCF\\ format\\ [500k\\ release]/chr{chrom}/ukb24310_c{chrom}_b{b_val}_v1.vcf.gz' \"\n","        cmd += f\"-iin='/Bulk/DRAGEN\\ WGS/DRAGEN\\ population\\ level\\ WGS\\ variants,\\ pVCF\\ format\\ [500k\\ release]/chr{chrom}/ukb24310_c{chrom}_b{b_val}_v1.vcf.gz.tbi' \"\n","        cmd += f\"-iin='{results_dir}/sample_ids.txt' \"\n","        cmd += f\"-icmd='bcftools view -O z -S sample_ids.txt ukb24310_c{chrom}_b{b_val}_v1.vcf.gz -o chr{chrom}_b{b_val}.vcf.gz' \"\n","        cmd += f\"--instance-type mem2_ssd1_v2_x2 \"\n","        cmd += f\"--destination '{projectid}:{results_dir}/01_pvcf_chunks'\"\n","\n","        result = subprocess.run(\n","            cmd, \n","            shell=True, \n","            stdout=subprocess.PIPE,\n","            stderr=subprocess.PIPE,\n","        )\n","\n","        if result.returncode != 0:\n","            print(f\"Error running command for chr = {chrom} b_val = {b_val}:\")\n","            print(result.stderr.decode(\"utf-8\"))\n"]},{"cell_type":"markdown","id":"5a97d311-e969-43c6-b2c8-eacff5b473d3","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["# Combine pVCF chunks"]},{"cell_type":"code","execution_count":null,"id":"4ed1146d-aaee-4cd5-b100-3b31c40cb52b","metadata":{},"outputs":[],"source":["\"\"\"\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------ PAUSE HERE UNTIL PREVIOUS STEP COMPLETES ------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","\"\"\""]},{"cell_type":"markdown","id":"69168913-442f-4d4a-9b4b-51c4870b2c02","metadata":{},"source":["## Use batches of 1000 files at first"]},{"cell_type":"code","execution_count":null,"id":"3b8935e7-d092-4672-b6ae-c84be5c28264","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["cmd = f\"dx run swiss-army-knife \"\n","icmd = \"-icmd='bcftools concat -O z \"\n","num_files = 0\n","\n","for _, row in df_b_vals.iterrows():\n","    start_bval = row[\"b_start\"]\n","    end_bval = row[\"b_end\"]\n","    chrom = row['chromosome']\n","    \n","    for b_val in range(start_bval, end_bval + 1):\n","        cmd += f\"-iin='{results_dir}/01_pvcf_chunks/chr{chrom}_b{b_val}.vcf.gz' \"\n","        icmd += f\"chr{chrom}_b{b_val}.vcf.gz \"\n","        num_files += 1\n","        if num_files % 1000 == 0:\n","            icmd += f\"-o concat_{num_files//1000}.vcf.gz' \"\n","            cmd += icmd\n","            cmd += f\"--instance-type mem2_ssd1_v2_x32 \"\n","            cmd += f\"--destination '{projectid}:{results_dir}/02_pvcf_concat'\"\n","            \n","            result = subprocess.run(\n","                cmd, \n","                shell=True, \n","                stdout=subprocess.PIPE,\n","                stderr=subprocess.PIPE,\n","            )\n","\n","            if result.returncode != 0:\n","                print(f\"Error running command:\")\n","                print(result.stderr.decode(\"utf-8\"))\n","            \n","            cmd = f\"dx run swiss-army-knife \"\n","            icmd = \"-icmd='bcftools concat -O z \" \n","\n","if num_files % 1000 != 0:\n","    icmd += f\"-o concat_{num_files//1000 + 1}.vcf.gz' \"\n","    cmd += icmd\n","    cmd += f\"--instance-type mem2_ssd1_v2_x32 \"\n","    cmd += f\"--destination '{projectid}:{results_dir}/02_pvcf_concat'\"\n","\n","    result = subprocess.run(\n","        cmd, \n","        shell=True, \n","        stdout=subprocess.PIPE,\n","        stderr=subprocess.PIPE,\n","    )\n","\n","    if result.returncode != 0:\n","        print(f\"Error running command:\")\n","        print(result.stderr.decode(\"utf-8\"))\n"]},{"cell_type":"markdown","id":"9d8f1bee-497b-476c-986a-4b2dfa794ea1","metadata":{},"source":["## Concat batches together"]},{"cell_type":"code","execution_count":null,"id":"9f50e2fc-f7b3-4224-9f10-41cccad002ae","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["cmd = f\"dx run swiss-army-knife \"\n","icmd = \"-icmd='bcftools concat -O z \"\n","for i in range(1, 13):\n","    cmd += f\"-iin='{results_dir}/02_pvcf_concat/concat_{i}.vcf.gz' \"\n","    icmd += f\"concat_{i}.vcf.gz \"\n","icmd += \"-o concat.vcf.gz' \"\n","cmd += icmd\n","cmd += f\"--instance-type mem2_ssd1_v2_x32 \"\n","cmd += f\"--destination '{projectid}:{results_dir}/02_pvcf_concat'\"\n","\n","result = subprocess.run(\n","    cmd, \n","    shell=True, \n","    stdout=subprocess.PIPE,\n","    stderr=subprocess.PIPE,\n",")\n","\n","if result.returncode != 0:\n","    print(f\"Error running command:\")\n","    print(result.stderr.decode(\"utf-8\"))\n"]},{"cell_type":"markdown","id":"434524be-1789-4c78-8e57-99f2e6de3b0e","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["# Normalize VCF before annotation"]},{"cell_type":"markdown","id":"cb462409-8441-4154-b27f-af2f460fa2fc","metadata":{"tags":[]},"source":["### Split multiallelic sites into biallelic records"]},{"cell_type":"code","execution_count":null,"id":"c0e635cb-8334-428d-800b-d535ab7c9641","metadata":{},"outputs":[],"source":["\"\"\"\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------ PAUSE HERE UNTIL PREVIOUS STEP COMPLETES ------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","\"\"\""]},{"cell_type":"code","execution_count":null,"id":"772711fe-c6f0-4905-8419-6e7eec331975","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["cmd = f\"dx run swiss-army-knife \"\n","cmd += f\"-iin='{results_dir}/02_pvcf_concat/concat.vcf.gz' \"\n","cmd += f\"-icmd='bcftools norm -m-both -O z -o biallelic.vcf.gz concat.vcf.gz' \"\n","cmd += f\"--instance-type mem2_ssd1_v2_x16 \"\n","cmd += f\"--destination '{projectid}:{results_dir}/03_pvcf_normalized'\"\n","\n","result = subprocess.run(\n","    cmd,\n","    shell=True,\n","    stdout=subprocess.PIPE,\n","    stderr=subprocess.PIPE,\n",")\n","\n","if result.returncode != 0:\n","    print(f\"Error running command:\")\n","    print(result.stderr.decode(\"utf-8\"))\n","    "]},{"cell_type":"markdown","id":"2b8f43a9-7b0c-4622-8651-5329845ec66c","metadata":{"tags":[]},"source":["### Left-align and normalize"]},{"cell_type":"code","execution_count":null,"id":"a960869a-67e1-464d-8ce8-b5b8cb666411","metadata":{},"outputs":[],"source":["\"\"\"\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------ PAUSE HERE UNTIL PREVIOUS STEP COMPLETES ------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","------------------------------------------------------------------------------------------------------\n","\"\"\""]},{"cell_type":"code","execution_count":null,"id":"cf1c6e99-309b-4ae0-9590-fa8191aef782","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["cmd = f\"dx run swiss-army-knife \"\n","cmd += f\"-iin='{results_dir}/03_pvcf_normalized/biallelic.vcf.gz' \"\n","cmd += f\"-iin='/data/Homo_sapiens_assembly38.fasta' \"\n","cmd += f\"-icmd='bcftools norm -f Homo_sapiens_assembly38.fasta -O z -o normalized.vcf.gz biallelic.vcf.gz' \"\n","cmd += f\"--instance-type mem2_ssd1_v2_x16 \"\n","cmd += f\"--destination '{projectid}:{results_dir}/03_pvcf_normalized'\"\n","\n","result = subprocess.run(\n","    cmd, \n","    shell=True, \n","    stdout=subprocess.PIPE,\n","    stderr=subprocess.PIPE,\n",")\n","\n","if result.returncode != 0:\n","    print(f\"Error running command:\")\n","    print(result.stderr.decode(\"utf-8\"))\n"]},{"cell_type":"markdown","id":"523f3e4e-743f-4605-b615-0c213ad9b6ee","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["# Check CRAM files for expansions"]},{"cell_type":"code","execution_count":null,"id":"30c644e3-a61b-4310-baa5-3238e901984f","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["%%bash\n","\n","dx download /data/Homo_sapiens_assembly38.fasta --overwrite\n","wget https://github.com/Illumina/ExpansionHunter/releases/download/v5.0.0/ExpansionHunter-v5.0.0-linux_x86_64.tar.gz\n","tar -xvzf ExpansionHunter-v5.0.0-linux_x86_64.tar.gz\n"]},{"cell_type":"code","execution_count":null,"id":"78816941-d2a7-4d82-abe6-39681edf3fdd","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["! dx download {results_dir}/chorea_cases.txt --overwrite\n","df_chorea = pd.read_csv(\"chorea_cases.txt\", sep=\"\\t\")\n","ids = df_chorea[\"ID\"].tolist()\n","\n","for iid in ids:\n","    ! dx download /Bulk/DRAGEN\\ WGS/Whole\\ genome\\ CRAM\\ files\\ \\(DRAGEN\\)\\ [500k\\ release]/{str(iid)[:2]}/{iid}_24048_0_0.dragen.cram --overwrite\n","    ! dx download /Bulk/DRAGEN\\ WGS/Whole\\ genome\\ CRAM\\ files\\ \\(DRAGEN\\)\\ [500k\\ release]/{str(iid)[:2]}/{iid}_24048_0_0.dragen.cram.crai --overwrite\n","    ! ExpansionHunter-v5.0.0-linux_x86_64/bin/ExpansionHunter --threads 8 --reads {iid}_24048_0_0.dragen.cram --reference Homo_sapiens_assembly38.fasta --variant-catalog ExpansionHunter-v5.0.0-linux_x86_64/variant_catalog/hg38/variant_catalog.json --output-prefix chorea_{iid}\n","    ! dx upload chorea_{iid}.vcf --path {results_dir}/00_expansions/chorea_{iid}.vcf\n","    ! rm {iid}_24048_0_0.dragen.cram\n","    ! rm {iid}_24048_0_0.dragen.cram.crai\n","    ! rm chorea_{iid}_realigned.bam\n","    ! rm chorea_{iid}.vcf\n","    ! rm chorea_{iid}.json\n"]},{"cell_type":"markdown","id":"3a67f581-c385-4642-88d0-90aa8ec2aec2","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["# Generate plink files"]},{"cell_type":"code","execution_count":null,"id":"8c0f344a-928a-4953-843d-753fa0ec802c","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["! dx download {results_dir}/chorea_cases.txt --overwrite\n","df_sex = pd.read_csv(f'chorea_cases.txt', sep=\"\\t\")\n","df_sex = df_sex[[\"ID\",\"GENETIC_SEX\"]]\n","df_sex[\"GENETIC_SEX\"] = df_sex[\"GENETIC_SEX\"].replace('Female', 2)\n","df_sex[\"GENETIC_SEX\"] = df_sex[\"GENETIC_SEX\"].replace('Male', 1)\n","df_sex.rename(columns={\"ID\":\"#IID\", \"GENETIC_SEX\":\"SEX\"}, inplace=True)\n","df_sex.to_csv(\"sex.txt\", sep=\"\\t\", index=False)\n","! dx upload sex.txt --path {results_dir}/sex.txt\n"]},{"cell_type":"code","execution_count":null,"id":"50fd769d-7f8b-448a-8615-59c445dce282","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["cmd = f\"dx run swiss-army-knife \"\n","cmd += f\"-iin='{results_dir}/03_pvcf_normalized/normalized.vcf.gz' \"\n","cmd += f\"-iin='{results_dir}/sex.txt' \"\n","cmd += f\"-icmd='plink2 --vcf normalized.vcf.gz --set-all-var-ids \\\"chr@:#:\\\\$r:\\\\$a\\\" --update-sex sex.txt --new-id-max-allele-len 1017 --make-pgen --out normalized' \"\n","cmd += f\"--instance-type mem2_ssd1_v2_x32 \"\n","cmd += f\"--destination '{projectid}:{results_dir}/04_plink'\"\n","\n","result = subprocess.run(\n","    cmd, \n","    shell=True, \n","    stdout=subprocess.PIPE,\n","    stderr=subprocess.PIPE,\n",")\n","\n","if result.returncode != 0:\n","    print(f\"Error running command:\")\n","    print(result.stderr.decode(\"utf-8\"))\n"]},{"cell_type":"markdown","id":"6a381f28-e318-45d9-a167-3fd40c98dc5b","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["# Annotation"]},{"cell_type":"markdown","id":"301fd58e-661c-49cc-858e-0f118d15395c","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["## Filter VCFs to only include a few participants"]},{"cell_type":"code","execution_count":null,"id":"a5696450-313d-4e80-bc55-00df2600aa68","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["cmd = f\"dx run swiss-army-knife \"\n","cmd += f\"-iin='{results_dir}/03_pvcf_normalized/normalized.vcf.gz' \"\n","cmd += f\"-icmd='bcftools view -O z -G normalized.vcf.gz -o annovar_input.vcf.gz' \"\n","cmd += f\"--instance-type mem2_ssd1_v2_x4 \"\n","cmd += f\"--destination '{projectid}:{results_dir}/05_annotated'\"\n","\n","result = subprocess.run(\n","    cmd, \n","    shell=True, \n","    stdout=subprocess.PIPE,\n","    stderr=subprocess.PIPE,\n",")\n","\n","if result.returncode != 0:\n","    print(f\"Error running command:\")\n","    print(result.stderr.decode(\"utf-8\"))\n"]},{"cell_type":"markdown","id":"50c8eaf9-1da4-49b8-ae56-890259085319","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["## Fetch Annovar libraries and reference genome data"]},{"cell_type":"code","execution_count":null,"id":"1a6d97d7-30e5-4aa4-b113-8a1a14129b8f","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["! wget http://www.openbioinformatics.org/annovar/download/0wgxR2rIVP/annovar.latest.tar.gz\n","! tar -xzf annovar.latest.tar.gz\n","! chmod a+x ./annovar/*.pl\n","! annovar/annotate_variation.pl -downdb -buildver hg38 -webfrom annovar refGene annovar/humandb/\n","! annovar/annotate_variation.pl -downdb -buildver hg38 -webfrom annovar avsnp151 annovar/humandb/\n","! annovar/annotate_variation.pl -downdb -buildver hg38 -webfrom annovar clinvar_20240917 annovar/humandb/\n","! annovar/annotate_variation.pl -downdb -buildver hg38 -webfrom annovar dbnsfp47a annovar/humandb/\n","! annovar/annotate_variation.pl -downdb -buildver hg38 -webfrom annovar dbnsfp47a_interpro annovar/humandb/\n","! annovar/annotate_variation.pl -downdb -buildver hg38 -webfrom annovar gnomad41_genome annovar/humandb/\n","! dx download /data/Homo_sapiens_assembly38.fasta --overwrite\n","! dx download {results_dir}/05_annotated/annovar_input.vcf.gz\n"]},{"cell_type":"markdown","id":"96fadfc5-73c5-4f79-ac67-b019a29f8322","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["## Perform annotation"]},{"cell_type":"code","execution_count":null,"id":"165edf19-d1ab-4751-89c7-9a126b700166","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["cmd = f\"annovar/table_annovar.pl annovar_input.vcf.gz annovar/humandb/ \"\n","cmd += f\"--buildver hg38 \"\n","cmd += f\"--thread 72 \"\n","cmd += f\"--remove \"\n","cmd += f\"--protocol refGene,avsnp151,clinvar_20240917,dbnsfp47a,dbnsfp47a_interpro,gnomad41_genome \"\n","cmd += f\"--operation g,f,f,f,f,f \"\n","cmd += f\"--nopolish \"\n","cmd += f\"--nastring . \"\n","cmd += f\"--out annotated \"\n","cmd += f\"--vcfinput \"\n","\n","result = subprocess.run(\n","    cmd, \n","    shell=True, \n","    stdout=subprocess.PIPE,\n","    stderr=subprocess.PIPE,\n",")\n","\n","if result.returncode != 0:\n","    print(f\"Error running command:\")\n","    print(result.stderr.decode(\"utf-8\"))\n"]},{"cell_type":"code","execution_count":null,"id":"b1f17ff0-f428-40f4-a1c2-2a407ffc6830","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["! mv annotated.hg38_multianno.txt annotated.txt\n","! dx upload annotated.txt --path {results_dir}/05_annotated/annotated.txt\n"]},{"cell_type":"markdown","id":"8d2e5906-4c04-4cde-afd0-59c0b0be3a69","metadata":{"jp-MarkdownHeadingCollapsed":true,"jupyter":{"outputs_hidden":true},"tags":[]},"source":["# Allele frequencies"]},{"cell_type":"code","execution_count":null,"id":"a7fa10b2-0df8-47c4-adb4-333545bcdf34","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["cmd = f\"dx run swiss-army-knife \"\n","cmd += f\"-iin='{results_dir}/04_plink/normalized.pgen' \"\n","cmd += f\"-iin='{results_dir}/04_plink/normalized.pvar' \"\n","cmd += f\"-iin='{results_dir}/04_plink/normalized.psam' \"\n","cmd += f\"-icmd='plink2 --pfile normalized --freq --out frequencies' \"\n","cmd += f\"--instance-type mem2_ssd1_v2_x16 \"\n","cmd += f\"--destination '{projectid}:{results_dir}/06_frequencies'\"\n","\n","result = subprocess.run(\n","    cmd, \n","    shell=True, \n","    stdout=subprocess.PIPE,\n","    stderr=subprocess.PIPE,\n",")\n","\n","if result.returncode != 0:\n","    print(f\"Error running command:\")\n","    print(result.stderr.decode(\"utf-8\"))\n"]},{"cell_type":"markdown","id":"7355d314-4504-45d7-a814-8a60821df159","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["# Zygosity"]},{"cell_type":"markdown","id":"fdf813fc-0f30-4a33-9d16-889cf89e92cf","metadata":{"jp-MarkdownHeadingCollapsed":true,"tags":[]},"source":["## Recode files"]},{"cell_type":"code","execution_count":null,"id":"e2b31b37-efdc-4dc0-aed5-b7f35f7ed577","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["for chrnum in list(range(1, 23)) + [\"X\",\"Y\",\"MT\"]:\n","    cmd = f\"dx run swiss-army-knife \"\n","    cmd += f\"-iin='{results_dir}/04_plink/normalized.pgen' \"\n","    cmd += f\"-iin='{results_dir}/04_plink/normalized.pvar' \"\n","    cmd += f\"-iin='{results_dir}/04_plink/normalized.psam' \"\n","    cmd += f\"-iin='{results_dir}/06_frequencies/frequencies.afreq' \"\n","    cmd += f\"-icmd='plink2 --pfile normalized --read-freq frequencies.afreq --chr {chrnum} --export A --out chr{chrnum}' \"\n","    cmd += f\"--instance-type mem2_ssd1_v2_x16 \"\n","    cmd += f\"--destination '{projectid}:{results_dir}/07_zygosity'\"\n","\n","    result = subprocess.run(\n","        cmd, \n","        shell=True, \n","        stdout=subprocess.PIPE,\n","        stderr=subprocess.PIPE,\n","    )\n","\n","    if result.returncode != 0:\n","        print(f\"Error running command:\")\n","        print(result.stderr.decode(\"utf-8\"))\n"]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.5"}},"nbformat":4,"nbformat_minor":5}
